### PR TITLE
feat: Add additional content

### PR DESCRIPTION
**Pull request by seebees**
Update the Readme with additional explanation and background

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,19 +1,186 @@
...
+## Persona: Developer
+
+For the developer writing the implementation
+Duvet maintains the punch list for you,
+you can focus on getting the best implementation for a given requirement.
+As you write the code you can ask:
+Dose this code satisfy this requirement?
+Does this test verify this behavior?
+If you need details,
+the document is right there with delicious cake-like background.
```
Comment (robin-aws) [inline README.md:81]: "cake-like" doesn't resolve in this context, expand on what you mean without using that term.

File: README.md [hunk 2]
```diff
@@ -1,19 +1,186 @@
...
+
+## Persona: Reviewer
+
+For the reviewer reading the code
+Duvet turns the design document into a map of the code.
+As you review the code the details are right there for you.
+You can jump right to the test and really focus on
+"Does this test convince me that the requirement MUST be met?"
+
+The reviewer can also updates the design.
```
Comment (robin-aws) [inline README.md:93]:
```suggestion
The reviewer can also update the design.
```

File: README.md [hunk 3]
```diff
@@ -1,19 +1,186 @@
...
+Running duvet on the new document
```
Comment (robin-aws) [inline README.md:94]:
```suggestion
When Duvet is run on the new document,
```

File: README.md [hunk 4]
```diff
@@ -1,19 +1,186 @@
...
+to both implement and test these new requirements.
 
-## Getting Started
+![](./img/reviewer.png)
+
+## Persona: Debugger  consumer?
+
+For the debugger coming to back to a project
+that they have not worked on in a long time,
+or are completely new to
```
Comment (robin-aws) [inline README.md:104]:
```suggestion
or are completely new to,
```

File: README.md [hunk 5]
```diff
@@ -1,19 +1,186 @@
...
+Duvet annotations create a network effect.
+These annotations help give context to code
```
Comment (robin-aws) [inline README.md:106]:
```suggestion
These annotations help give context to code,
```

File: README.md [hunk 6]
```diff
@@ -1,19 +1,186 @@
...
+especially if the code is not your preferred language.
+They link back to the design document
+to give background and more information.
+Since these annotations are bidirectional,
+are waypoints that let you move between
```
Comment (robin-aws) [inline README.md:111]:
```suggestion
they act as waypoints that let you move between
```

File: README.md [hunk 7]
```diff
@@ -1,19 +1,186 @@
...
+the more abstract design document
+and the detailed implementation.
+
+![](./img/debugger.png)
+
+## When do you stop adding tests?
+
+But now we are done right?
+Every requirement has an implementations and a tests.
```
Comment (robin-aws) [inline README.md:120]:
```suggestion
Every requirement has an implementation and a test.
```

File: README.md [hunk 8]
```diff
@@ -1,19 +1,186 @@
...
+They all have been reviewed.
+Duvet can only see that you TRIED to test.
+How does the reviewer know you tested ENOUGH?
+
+Duvet is a mechanism to connect the requirement to the source,
+but it leaves the quality and completeness of these links to humans.
+A tests makes sure that something is true once,
+it does not make sure that something is *always* true.
```
Comment (robin-aws) [inline README.md:128]:
```suggestion
A tests makes sure that something is true for one particular set of inputs,
it does not make sure that something is *always* true.
```

File: README.md [hunk 9]
```diff
@@ -1,19 +1,186 @@
...
+Todays edge case it tomorrow's issue.
+
+## Dafny
+
+Dafny is a *verification-aware* programming language.
+
+Formal verification let's us translate the Duvet annotation
```
Comment (robin-aws) [inline README.md:135]:
```suggestion
Formal verification lets us translate the Duvet annotation
```

File: README.md [hunk 10]
```diff
@@ -1,19 +1,186 @@
...
+This is like writing a test, but you are testing every value all at once
+statically before you run any code.
+
+Dafny uses formal logic to prove specifications.
+It doesn't forget about edge cases
+because it checks every case.
+
+Since is very useful to have verified code,
+but only if you can run it
+Dafny compiles to many other languages:
```
Comment (robin-aws) [inline README.md:148]:
```suggestion
Since is useful to have verified code,
but much more useful if you can run it,
Dafny compiles to many other languages:
```

Just to avoid implying that modeling and proving some other code over there is not useful. :)

File: README.md [hunk 11]
```diff
@@ -1,19 +1,186 @@
...
+
+The reviewer can also updates the design.
+Running duvet on the new document
+the developer is now on the hook
+to both implement and test these new requirements.
 
-## Getting Started
+![](./img/reviewer.png)
+
+## Persona: Debugger  consumer?
```
Comment (robin-aws) [inline README.md:100]:
```suggestion
## Persona: Debugger
```

Or was that meant as a question to the PR reviewer?

File: README.md [hunk 12]
```diff
@@ -1,19 +1,186 @@
...
+
+Since is very useful to have verified code,
+but only if you can run it
+Dafny compiles to many other languages:
+Java, C#, JavaScript, Go, and more on the way!
+
+![](./img/dafny.png)
+
+## Mechanisms!
+
```
Comment (robin-aws) [inline README.md:154]:
```suggestion
```

I think this was useful as a slide header but not here.

File: README.md [hunk 13]
```diff
@@ -1,19 +1,186 @@
...
+Putting this together we have a strong chain
+from the designer's intent right to what code is executing.
+
+Now we have two mechanisms.
+First Duvet makes us write a specification for every requirement.
+Then Dafny proves that our implementation matches our specification.
+
+![](./img/mechanisms.png)
```
Comment (robin-aws) [inline README.md:162]: Couple of jagged arrows here, were those still in the latest deck?

File: README.md [hunk 14]
```diff
@@ -1,19 +1,186 @@
...
+## Let’s start!
 
 Just follow along with [the workshop instructions](./instructions/steps.md) right here on GitHub!
 
+## To learn more
+
+* [Dafny](http://dafny.org/), a programming language that formally verifies your implementation matches your specification
+* [Duvet](https://github.com/awslabs/duvet), a code quality tool for measuring how well your human-readable specification is covered in code
+* AWS Encryption SDK in Dafny: [https://github\.com/aws/aws\-encryption\-sdk\-dafny](https://github.com/aws/aws-encryption-sdk-dafny)
+* s2n: [https://github\.com/aws/s2n\-tls](https://github.com/aws/s2n-tls)
```
Comment (robin-aws) [inline README.md:182]: Yup, definitely mention these towards the top somewhere.

File: README.md [hunk 15]
```diff
@@ -1,19 +1,186 @@
...
+
+Let's keep it positive.
+Correct software improves
+
+* Operations -
+  Your software performs as you expect, more uptime.
+* Evolution -
+  You know what needs to change, faster development.
+* Security -
+  You can reason about security properties, better sleep.
```
[the remark this answers is not in the document]
Comment (seebees) [reply README.md:23]: I'm inclined to leave it.
Since the point is exactly that this "correctness is important"
is not always so obvious.
Empirically, since everything is not formally verified with tests and a specification :)

File: README.md [hunk 16]
```diff
@@ -1,19 +1,186 @@
...
+
+## How hard can it be?
+
+The problem with correctness is it is an external property.
+You can read code, but you can not know by the code alone that it is correct.
```
[the remark this answers is not in the document]
Comment (seebees) [reply README.md:28]: took

File: README.md (hunk 14 above)
Comment (seebees) [reply README.md:182]: moved

**Review by robin-aws — APPROVED** [review]
🚀
